pkg/build/addimage/buildcontext.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addimage
+
+import (
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/kind/pkg/build/addimage/internal/container/docker"
+	"sigs.k8s.io/kind/pkg/errors"
+	"sigs.k8s.io/kind/pkg/exec"
+	"sigs.k8s.io/kind/pkg/fs"
+	"sigs.k8s.io/kind/pkg/log"
+)
+
+// buildContext is used to build the kind node image, and contains
+// build configuration
+type buildContext struct {
+	// option fields
+	image            string
+	baseImage        string
+	additionalImages []string
+	logger           log.Logger
+	arch             string
+}
+
+// Build builds the cluster node image, the sourcedir must be set on
+// the buildContext
+func (c *buildContext) Build() (err error) {
+	return c.addImages()
+}
+
+func (c *buildContext) addImages() error {
+
+	c.logger.V(0).Info("Starting to add images to base image")
+	// pull images to local docker
+	for _, imageName := range c.additionalImages {
+		// Check to see if the image exists and if not pull it
+		_, err := docker.ImageID(imageName)
+		if err != nil {
+			err = docker.Pull(c.logger, imageName, dockerBuildOsAndArch(c.arch), 3)
+			if err != nil {
+				c.logger.Errorf("Add image build Failed! Failed to pull image %v: %v", imageName, err)
+			}
+		}
```

Comment:
nit, this reads better unnesting
```suggestion
              // continue if image already exist, no need to pull
		if err == nil {
		        continue
		}
              err = docker.Pull(c.logger, imageName, dockerBuildOsAndArch(c.arch), 3)
		if err != nil {
			c.logger.Errorf("Add image build Failed! Failed to pull image %v: %v", imageName, err)
		}
	}
```